Add basic gameplay leaderboard display

peppy opened the pull request:
This is by no means the final product, but is a stepping stone towards where we want to be. It also completes the "stable parity" requirement at least.

"Why is this not a customisable HUD component?"

Because the implementation changes based on context (multiplayer vs solo currently, but may also support playlists at a later point), it makes it hard to add position flexibility. Especially since in multiplayer we want it to always display.

For now it makes more sense to get the flow in place, then figure out how to allow customisation of positioning at a later point if/when there is demand.

There's also some further work I want to do on the leaderboard itself (positional updates should be debounced differently to prioritise immediate response if the leaderboard hasn't updated recently, for example) but I'll make those changes separately.

Closes #18879.

- [ ] Depends on #20389 (for test to show up properly).

frenzibyte committed:
Add simplified `SoloGameplayLeaderboard` implementation

Note that this doesn't support waiting until leaderboard finishes fetching scores.

frenzibyte committed:
Add test coverage

peppy committed:
Track local `Player.Score`'s user rather than using `APIProvider`

peppy committed:
Make constructor `protected` in base implementation

peppy committed:
Move leaderboard implementation to `Player` itself

peppy committed:
Allow `DrawableAvatar` to accept an `IUser` for now

peppy committed:
Change `Leaderboard` to use `IUser` instead of `APIUser`

peppy committed:
Refactor `SoloGameplayLeaderboard` to not read scores via DI

Also allows updating scores if they arrive late.

peppy committed:
Add back DI leaderboard retrieval via bindable pathway

peppy committed:
Ensure gameplay leaderboard hides with rest of HUD when it should

peppy committed:
Move `LeaderboardFlow` to `HUDOverlay` to share positioning logic

peppy committed:
Fix leaderboard positioning logic to actually consider elements on the left side

peppy committed:
Fix oversight when making interface implementation explicit

peppy committed:
Fix back-to-front condition on leaderboard always show configuration

Feodor0090 commented:
Mod multipliers?
![изображение](https://user-images.githubusercontent.com/53872073/190015941-75c03341-c3fb-4c87-b370-ca3e2f0f0087.png)
![osu_2022-09-14_00-50-15](https://user-images.githubusercontent.com/53872073/190015980-125eac38-cc9f-455a-8c48-b307c486666d.png)
The 4th score displays without x0.1 AP multiplier here.

peppy committed:
Fix displayed scores in gameplay leaderboard not tracking display mode changes

peppy committed:
Fix gameplay leaderboard not being sorted correctly in tie situations

peppy commented:
@Feodor0090 thanks for testing. i've attempted to resolve your issues and it looks to be better in local testing. can you re-test with your test scenarios to confirm things are better?

Feodor0090 commented:
> can you re-test with your test scenarios to confirm things are better?

Both issues are fixed and i could not find any other ways to break it yet.

smoogipoo commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -22,29 +22,40 @@
 namespace osu.Game.Screens.Select
 {
-    public class PlaySongSelect : SongSelect
+    public class PlaySongSelect : SongSelect, ILeaderboardScoreSource
```
Is this still required? I find this implementation exceedingly weird. It's forwarding the score source to `playBeatmapDetailArea.Leaderboard.Scores`, but `BeatmapLeaderboard` is already an `ILeaderboardScoreSource`?

And, part of the bigger picture, I am very confused about this `ILeaderboardScoreSource` stuff. Why is a leaderboard a score source and/or why is the leaderboard not the _single_ score source ever? It's like, `SoloGameplayLeaderboard` binds _outwards_ now but still has a perfectly accessible `.Add()` method.

The beatmap leaderboard pulls things from the database or online, the multiplayer leaderboard opens spectator channels and populates its own scores/manages their own states, why can't solo player do this too?

smoogipoo replied:
Is it just that `PlaySongSelect` is an `ILeaderboardScoreSource` which is cached for when entering `Player` so that the leaderboard can then look it up?

It's just too much. The only reason that I can come up with this is to save on API requests? Which is fine, but in the process of doing that the code quality is completely destroyed and logic now goes all around the place and is unreadable.

peppy replied:
It's just for the caching, yes.

It means that whatever is displayed at song select is used during gameplay (similar to stable).

The whole flow will need to change in the future (as I want to be able to dynamically load in/out scores using ongoing web requests), but the idea was to keep things simple in this implementation.

smoogipoo replied:
I don't feel good merging this. I suggest refactoring it in this PR and/or a pre-requisite PR that has a saner structure.

peppy replied:
I'll see what I can do.

Would you prefer something like a bindable exposed on `Player` which is bound to by `PlaySongSelect`? Is the main issue the DI interface?

smoogipoo replied:
The main issue I have is the DI interface, yes.

However it's a question of what you're actually trying to achieve here:
- Is it about retrieval type (i.e. which tab is selected in song select) and if so why can't that be re-queried again? As you mentioned, it's going to need some extra logic anyway to re-query the intermediate scores as the player progresses up the rankings so there's little point (in my mind) in taking on song select's rankings.  
  - The selected tab is already stored to the `OsuConfigManager` (`BeatmapDetailTab`+`BeatmapDetailModsFilter`) so there shouldn't be any problems.
- Is it about saving on web requests? If so, perhaps a separate component is better fit for this? For example, what happens when a user replays a map, is the client going to re-query all the intermediate scores for the second play? If it does, is there any point on saving web requests - I assume we want the most recent data (especially for newly-ranked beatmaps).

peppy replied:
If you look back at the history of this PR, the component was made in a way that would always theoretically work with `PlaylistPlayer`. As long as a parent has a score source, it would just-work.

I removed the playlist portion because frenzi made the incorrect assumption about `AggregateScore`s being per-beatmap.

> * Is it about retrieval type (i.e. which tab is selected in song select) and if so why can't that be re-queried again?

Yes, things like the mods filter etc. would need to be communicated if you wanted to do a second redundant lookup.

>   * The selected tab is already stored to the `OsuConfigManager` (`BeatmapDetailTab`+`BeatmapDetailModsFilter`) so there shouldn't be any problems.

> * Is it about saving on web requests?

Yeah, I wouldn't want to query the same thing twice at two consecutive screens, for sure.

smoogipoo replied:
Well `PlaylistPlayer` is also a child of `Playlists` (apparently what it's called, no `Screen` suffix...) and _not_ the `PlaylistsRoomSubscren`. So I don't believe it would work there unless I'm missing something.

Are you suggesting that `Playlists` implement this new interface, and then the subscreen uses scores DI'd from that? And then `Playlists` somehow knows about all the interactions going on such as what if the leaderboard inside the subscreen is scrolled - should more scores come up, etc.

peppy replied:
The playlist implementation was equally weird with redirection, yes. I'm not saying that would return (currently playlist only retrieves scores when entering the results screen) – it needs more thought for sure.

I do think there's value in having a shared data source between song select and player though (in terms of streamlining requests and realm overhead), rather than requesting at each location. I'll give a bindable source a try and see how things go.

peppy replied:
I've changed the flow, see if it works any better to your eye.

peppy committed:
Merge branch 'master' into gameplay-leaderboards

peppy committed:
Provide scores directly to `Player` instance rather than relying on DI

frenzibyte commented on osu.Game/Screens/Play/HUDOverlay.cs:
```diff
@@ -177,22 +189,36 @@ protected override void Update()
             // LINQ cast can be removed when IDrawable interface includes Anchor / RelativeSizeAxes.
             foreach (var element in mainComponents.Components.Cast<Drawable>())
             {
                 // for now align top-right components with the bottom-edge of the lowest top-anchored hud element.
```
I guess this should be reworded to:
```suggestion
                // for now align top components with the bottom-edge of the lowest top-anchored hud element.
```
Assuming the condition was updated to have the leaderboard position itself below top-left components as well?

frenzibyte commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -820,6 +823,40 @@ protected override bool OnScroll(ScrollEvent e)
+            });
+        }
+
+        public readonly BindableList<ScoreInfo> LeaderboardScores = new BindableList<ScoreInfo>();
```
I would move this to `SoloPlayer` and `ReplayPlayer` rather than place it here, since not all players are based off a leaderboard with finalised scores.

frenzibyte commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -820,6 +823,40 @@ protected override bool OnScroll(ScrollEvent e)
+
+        public readonly BindableList<ScoreInfo> LeaderboardScores = new BindableList<ScoreInfo>();
+
+        protected virtual GameplayLeaderboard CreateGameplayLeaderboard() => new SoloGameplayLeaderboard(Score.ScoreInfo.User)
```
In addition to my comment above about `LeaderboardScores` bindable, I would make this abstract and override in `SoloPlayer`/`ReplayPlayer`. `Player` shouldn't create a "solo" component.

peppy replied:
yes well. i was avoiding having to do this in every implementation except multiplayer. i guess it's fine to do so though.

frenzibyte commented on osu.Game/Screens/Select/PlaySongSelect.cs:
```diff
@@ -100,14 +110,18 @@ protected override bool OnStart()
+                else
+                    player = new SoloPlayer();
+
+                ((IBindableList<ScoreInfo>)player.LeaderboardScores).BindTo(playBeatmapDetailArea.Leaderboard.Scores);
```
I think it wouldn't be a problem changing `LeaderboardScores` to `IBindableList<ScoreInfo>`? Just to get rid of this ugly cast.

peppy replied:
We generally only use `IBindable` for read-only usages, so I'm not sure.

frenzibyte replied:
Well on `SoloGameplayLeaderboard`, it shouldn't be mutating `LeaderboardScores` I think? It's only using it as a source for score cards.

peppy replied:
After moving the leaderboard more locally, this is no longer an issue.

frenzibyte commented:
Beginning gameplay with leaderboard in "fetching" state doesn't seem to actually populate the scores.

https://user-images.githubusercontent.com/22781491/190864772-84a3aed8-7cfe-4424-bf2b-219517a31565.mp4

(notice how request finished succesfully during gameplay but no scores are shown still)

peppy committed:
Move implementation to individual classes

peppy committed:
Update comment to not mention horizontal plane

peppy committed:
Ensure `Leaderboard.Scores` is updated immediately via request flow

peppy commented:
> Beginning gameplay with leaderboard in "fetching" state doesn't seem to actually populate the scores.

Nice catch. I've fixed this by moving the point of schedule to a more amicable location. Can be tested using something like this:

```diff
diff --git a/osu.Game/Online/API/APIAccess.cs b/osu.Game/Online/API/APIAccess.cs
index a0c8e0d555..8e35f474c8 100644
--- a/osu.Game/Online/API/APIAccess.cs
+++ b/osu.Game/Online/API/APIAccess.cs
@@ -342,6 +342,9 @@ private bool handleRequest(APIRequest req)
         {
             try
             {
+                if (req is GetScoresRequest)
+                    Thread.Sleep(10000);
+
                 req.Perform(this);
 
                 if (req.CompletionState != APIRequestCompletionState.Completed)

```

frenzibyte commented on osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs:
```diff
@@ -0,0 +1,73 @@
+            if (!Scores.Any())
+                return;
+
+            ILeaderboardScore local = Add(trackingUser, true);
+
+            local.TotalScore.BindTarget = scoreProcessor.TotalScore;
+            local.Accuracy.BindTarget = scoreProcessor.Accuracy;
+            local.Combo.BindTarget = scoreProcessor.Combo;
+
+            // Local score should always show lower than any existing scores in cases of ties.
+            local.DisplayOrder.Value = long.MaxValue;
+
+            foreach (var s in Scores)
```
Once player loads, the local player is initially placed at the top of the leaderboard then sorted back down a moment later. I think adding it after existing scores would work better.

![CleanShot 2022-09-20 at 23 39 13@2x](https://user-images.githubusercontent.com/22781491/191360327-53d0d901-09cc-4000-8adb-64f610832b46.png)

peppy replied:
I actually thought this was an issue with the base leaderboard implementation and was going to leave for a follow-up fix, but this makes sense.

frenzibyte commented on osu.Game/Screens/Play/HUD/SoloGameplayLeaderboard.cs:
```diff
@@ -0,0 +1,73 @@
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            Scores.BindCollectionChanged((_, _) => Scheduler.AddOnce(showScores), true);
```
When finishing gameplay, the user's submitted online score seems to be added to the `Scores` list, which makes it appear on the leaderboard:

https://user-images.githubusercontent.com/22781491/191360700-d526ed6b-148c-4089-bd09-d3efbd4114d9.mp4

I think that could be fixed by checking against the score ID, similar to how it's done on results screens.

peppy replied:
This is a bit unfortunate, as it was avoided by having the schedule in place previously.

I've fixed this by unbinding the scores, rather than trying any kind of ID matching (sounds awkward).

frenzibyte commented on osu.Game/Online/Leaderboards/Leaderboard.cs:
```diff
@@ -39,7 +39,9 @@ public abstract class Leaderboard<TScope, TScoreInfo> : CompositeDrawable
         /// <summary>
         /// The currently displayed scores.
         /// </summary>
-        public IEnumerable<TScoreInfo> Scores => scores;
+        public IBindableList<TScoreInfo> Scores => scores;
+
+        private readonly BindableList<TScoreInfo> scores = new BindableList<TScoreInfo>();
```
This doesn't seem to reset when selecting another beatmap while on "Details" section, causing the gameplay session to show a leaderboard with scores from another beatmap.

peppy replied:
I've fixed this by clearing the old scores for now.

I think in the future we'd either want the leaderboard to be always present at song select (ie. with the new design), or the retrieval should be done by its own component (as @smoogipoo touched on earlier).

frenzibyte reviewed (commented):
Have play-tested this and it looks to work pretty fine except for the mentioned issues.

peppy committed:
Add local score after existing scores to avoid initial re-sort

peppy committed:
Fix solo leaderboard seeing imported score via realm subscription flow

peppy committed:
Add failing test showing layout failure in gameplay leaderboard

peppy committed:
Add extra margin space to flow equal to height of leaderboard

This ensures the content is always on screen, but also accounts for the
fact that scroll operations without animation were actually forcing the
local score to a location it can't usually reside at.

Basically, the local score was in the scroll extension region (due to always trying
to scroll the local player to the middle of the display, but there being
no other content below the local player to scroll up by).

peppy committed:
Animate scroll for a better visual experience

peppy committed:
Merge branch 'fix-gameplay-leaderboard-layout' into gameplay-leaderboards

peppy committed:
Order test scores more correctly

peppy committed:
Fix outdated scores being held by `BeatmapLeaderboard` after external beatmap switch

peppy commented:
I've addressed all the new issues found. May look to add some more test coverage, but still welcome further review on this (can probably go ahead without more tests TBH).

smoogipoo committed:
Clean up some checks

smoogipoo commented:
Crash on entering song select on dev server:
```
Unhandled exception. NUnit.Framework.AssertionException: : 
   at osu.Framework.Logging.ThrowingTraceListener.Fail(String message1, String message2)
   at System.Diagnostics.TraceInternal.Fail(String message, String detailMessage)
   at System.Diagnostics.TraceInternal.TraceProvider.Fail(String message, String detailMessage)
   at System.Diagnostics.Debug.Fail(String message, String detailMessage)
   at osu.Game.Online.Leaderboards.Leaderboard`2.SetErrorState(LeaderboardState state) in /home/smgi/Repos/osu/osu.Game/Online/Leaderboards/Leaderboard.cs:line 172
   at osu.Game.Screens.Select.Leaderboards.BeatmapLeaderboard.FetchScores(CancellationToken cancellationToken) in /home/smgi/Repos/osu/osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs:line 112
   at osu.Game.Online.Leaderboards.Leaderboard`2.refetchScores() in /home/smgi/Repos/osu/osu.Game/Online/Leaderboards/Leaderboard.cs:line 229
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
```

peppy committed:
Fix reversed checks

peppy commented:
Fixed >_>
